### ISSUE TITLE
Data backup fails for some apps

### ISSUE THREAD

**Issue by TiredRobot** [machiav3lli/oandbackupx#84]
Out of 61 apps I tried to backup, 5 of them failed.

Two of them gave me the following errors:
> BackupFailedException: cp: /data/user/0/com.taxis99/files/hotpatch-dexopt: No such file or directory
BackupFailedException: cp: /data/user/0/org.mozilla.firefox/files/mozilla/vat1yrpl.default/lock: No such file or directory

Both of these are 0 byte files with the attributes `lrwxrwxrwx`, so maybe oabx doesn't handle the `l` attribute properly?

The other three apps gave me a different error:
>BackupFailedException: cp: /storage/emulated/0/OABX/br.com.jamesdelivery.android.james/data/files/default.realm.management/access_control.pick_writer.cv: Invalid argument
BackupFailedException: cp: /storage/emulated/0/OABX/com.moez.QKSMS/data/files/default.realm.management/access_control.pick_writer.cv: Invalid argument
BackupFailedException: cp: /storage/emulated/0/OABX/com.opera.mini.native/data/files/turboproxy: Invalid argument

These are also 0 byte files. The first two have the attributes `prwx------` and the third one `srwx------`.

Also (and maybe this should be a separate issue), after a batch apk+data backup failed for these apps:

- Trying to backup the APK of one of the first two apps I mentioned makes oabx reload the app list and not be able to read user apps properly (?) until I restart it. 
- Trying to backup the APK of any of the other three apps crashes oabx with the following error:
> msg: java.lang.NullPointerException
stacktrace: java.lang.NullPointerException
	at java.io.File.<init>(File.java:283)
	at com.machiav3lli.backup.handler.action.BackupAppAction.removePackageBackup(BackupAppAction.java:266)
	at com.machiav3lli.backup.handler.action.BackupAppAction.cleanBackup(BackupAppAction.java:70)
	at com.machiav3lli.backup.handler.BackupRestoreHelper.backup(BackupRestoreHelper.java:40)
	at com.machiav3lli.backup.activities.BatchActivityX.doAction(BatchActivityX.java:209)
	at com.machiav3lli.backup.activities.BatchActivityX.lambda$onConfirmed$6$BatchActivityX(BatchActivityX.java:177)
	at com.machiav3lli.backup.activities.-$$Lambda$BatchActivityX$6vwFSSS37VmuQbaM9AdoAx4lavY.run(Unknown Source:4)
	at java.lang.Thread.run(Thread.java:919)

I was able to fix those by deleting their respective folders in OABX's backup folder.

**Comment by machiav3lli** [comment]
Those crashes are caused by not being able to copy pipes/ symlinks. But the crashing is a new thing. Do you have "verify apps over usb" in developer settings off or on?

**Comment by TiredRobot** [comment]
Off.

**Comment by Micha-Btz** [comment]
I have the same error while backing up on fennec from f-droid store. 
`BackupFailedException: cp: /data/user/0/org.mozilla.fennec_fdroid/files/mozilla/$$$$$$.default/lock: No such file or directory`

Would it be possible to exclude some files from apps, like rsync --exclude?

**Comment by Tiefkuehlpizze** [comment]
I'm not sure, if that's really what you would desire. This could bring your app into an inconsistent state where it's only partly initialized. The issues are known and it's being worked on, but it still needs some time because the changes are huge.

Thanks for your idea, but try to imagine, how you would expect to set this exclude filter. You might notice a dilemma a dev would run into. :smiley:

**Comment by Micha-Btz** [comment]
Yes, you are right. But I don't think on a per app solution. Maybe a general leaving out of symlinks would be an idea.

**Comment by Tiefkuehlpizze** [comment]
Maybe, but I would rather having my backup to fail than having a corrupt backup.

When Nougat (Android 7) gets officially dropped (and I'm 99.9% sure, it will) the app is finally able to distinguish special files such as pipes and symlinks from "normal" files and directories. So there's some light.

**Comment by Micha-Btz** [comment]
Ok, for me it crashed also. I add a logcat.

[logcat_08-10-2020_10-24-06.txt](https://github.com/machiav3lli/oandbackupx/files/5049781/logcat_08-10-2020_10-24-06.txt)

**Comment by 07416** [comment]
I have the same issue with `.lock` file with Firefox Fenix (beta), only the APK gets backed up.

**Comment by opusforlife2** [comment]
Backup and restore of Fenix _did_ work on version 1.2 for me, though. Same with XDA Labs. One has a symlink, and the other has a named pipe. Maybe whatever worked then could be used as a workaround? Unless it would be more trouble than it's worth, of course.

**Comment by Tiefkuehlpizze** [comment]
For clarification: Mozilla's lock file is a symlink. It vanishes when Firefox is exits (not killed).
```generic_x86:/data/data/org.mozilla.fenix/files/mozilla/p6xfhavt.default # ls -l lock 
lrwxrwxrwx 1 u0_a74 u0_a74 15 2020-08-10 16:54 lock -> 127.0.0.1:+6059
```

The old version of oandbackup(x) used copying and zip for backup and restore.
I won't test it, but since sdcardfs (FAT32) does not support symlinks, you can expect, that those aren't present in your zip file backup.
The error messages you receive "No such file or directory" comes from an invalid symlink. Firefox seems to transfer some hostname + port via the symlink target. I'll never be valid.
The other error "Invalid Argument" comes, because you cannot simply read the contents of a pipe. And cp seems to do that for copying.

The difficult part here is, to get all the meta information and process the files accordingly. On top, this must be done using a root subprocess such as `ls` or `stat`, because the app itself is running as normal userspace app.

I don't like doing this, but it seems, that we need to transfer (parse) the output of `ls` with all the metadata into the app, so we now, that there are special types.
This is again a major change in backup/restore mechanism since we're not dealing not really dealing with accessible paths anymore. It's more like: "Here's a path. It's a file. su-read it and add it to the tar archive" for backup.
It makes processing a bit more complicated, but the good thing is, that it directly streams the file from the "private" storage of the app (name it _data_ ) into the archive and this means on it's way it's being compressed and encrypted.

This also eliminates the copying issue, that the potentially sensitive information is temporary stored in plain text on the storage before it's compress, encrypted and added to the archive. Even though su-reading the file with libsu might give a penalty in CPU cycles used to do that, it avoids copying stuff around completely. This saves flash writing cycles and time at cost at some CPU load.

I hope, I could make some of my thoughts transparent and understandable and of course describe the problem with illegally "stealing" an app's data (at least according to Android's Security principles :smile: ). Google of course wants you to use their backup provider, but this is a completely different story because it's being privileged by the Android system or manufacturer to do nasty things. I don't think OandBackupX is aiming to be an alternative to that right now, because then it would have to be installed to the rom like OpenGapps.

**Comment by opusforlife2** [comment]
I wasn't 100% sure, so I just rechecked. Quitting either Fenix or Fennec (and optionally removing it from Recents) doesn't remove that lock file. What's wrong here? If this can be solved, then at least Fenix and Fennec backups can be unblocked.

> I won't test it, but since sdcardfs (FAT32) does not support symlinks, you can expect, that those aren't present in your zip file backup.

Well, apparently this file isn't all that important, because I've restored from such a backup several times and it's been fine. At least for Fenix.

**Comment by machiav3lli** [comment]
The 1.2.0 still had the oab-utils which managed the commands of copying and deleting on a system level(using su). The lock file doesn't have anything with the the app being up or killed, it's more of a centralisation method for mozilla browsers managing common resources. Eventually, we'll be trying to support symlinks in the all possible ways, except resulting to a secondary utility like oab-utils. Thanks for your active feedback.

**Comment by Smojo** [comment]
I can agree that "Fennec F-Droid" backup is not working.

Not sure if it is the right place, but I just post a list of apps here I currently have on my backlist (to be able to run a batch/schedued backup of all [other] apps+data).

Most of them just lead to a crash of OABX without any message in logcat.

User apps:
- Amaze
- Mein Vodafone*
- Aurora Store
- Cythara
- Termux -> logcat error-messages available
- Komoot* -> logcat error-messages available
- Fennec F-Droid

System apps / Special backups:
- Accounts 
- Android System -> logcat error-messages available

'* these apps (marekd with a star) were downloaded via Aurora Store so I they are from Google Play Store 

All the others are installed via F-Droid. 
I have also some other apps installed via Aurora. 
So several apps, small and big ones, complex and simple ones.
If needed I can send the logcat output for the apps which produce errors.
Ones an app leads to a crash it frequently writes a message like:
`[ <timestamp> 29510:29892 E/OAndBackupX.LogFile ]
<packagename>: error while reading logfile: org.json.JSONException: Value java.io.FileNotFoundException of type java.lang.String cannot be converted to JSONObject`
to logcat.

**Comment by 07416** [comment]
^https://play.google.com/store/apps/details?id=com.funanduseful.earlybirdalarm&hl=en also fails. The app has an internal backup feature, but data is uploaded to their server (blocked in firewall on my phone).

**Comment by Smojo** [comment]
Works in the meanwhile:
- Amaze
- Mein Vodafone
- Aurora Store
- Cythara
- Termux (works, after I reinstalled it but data bkp takes disproportionately long)

Btw.: I always backed up both. APK+Data
Just tested some of the apps apk and data backup separatly. As the issue title states the just fail during data backup.
- Komoot fails with:
`BackupFailedException: cp: /storage/emulated/0/OABX/de.komoot.android/data/files/default-realm.note invalid argument`

**Comment by Tiefkuehlpizze** [comment]
Termux is a quite special case. It's extracting a rootfs in it's data directory and this means it creates a huge amount of small files. Unfortunately it'll always need a big amount of time.
Maybe it can be improved by offering some alternative way of processing, using the cache directory and toybox to create a intermediate tar file, so small files don't have to be processed through the Java code.

Just an idea. Sounds like a dirty workaround

**Comment by nyanpasu64** [comment]
When opening OABX, I get a JSONException in logcat:

```
11-01 22:47:06.218 13202 13245 E OAndBackupX.LogFile: app.zxtune: error while reading logfile: org.json.JSONException: Value java.io.FileNotFoundException of type java.lang.String cannot be converted to JSONObject
```

Backing up all user apps fails with a NullPointerException at removePackageBackup (for the same corrupt app, zxtune), which aborted the backup process and kicked me back to the main app window.

Deleting the corrupt ZXTune backup in the GUI (it shows up as 1969, due to Unix time 0 and time zones) fixed the issue and allowed me to backup all apps.

Glodroid 0.4.0 on PinePhone.

**Comment by 07416** [comment]
Backup works in v4.0.0 without errors in at least beta (`version 84.0.0-beta.2`). I haven't yet tried restoring.
